crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1307,6 +1307,48 @@ impl<'src> Lexer<'src> {
         }
     }
 
+    /// Re-lex the current token in the context of a logical line.
+    ///
+    /// Returns a boolean indicating that whether the new current token is different than the
+    /// previous current token. This also means that the current position of the lexer has changed
+    /// and the caller is responsible for updating it's state accordingly.
+    ///
+    /// This method is a no-op if the lexer isn't in a parenthesized context.
+    pub(crate) fn re_lex_logical_token(&mut self) -> bool {
+        if self.nesting == 0 {
+            return false;
+        }
+
+        // Reduce the nesting level because the parser recovered from an error inside list parsing.
+        self.nesting -= 1;
```

Comment:
Should the below logic run when the nesting level remains greater than one? I woudl expect that it shouldn't change the lexed token because newlines must still be lexed as `NonLogicalNewline`.

Reply:
That's a good point. It shouldn't really be required but in case of nested list parsing like for example:

```py
if call(foo, [a, b
    def bar():
        pass
```

If we don't move the lexer back to the newline character after `b`, the "expected `]`" error will be at the `def` token:

```
Syntax Error: Expected ']', found 'def' at byte range 23..26
  |
1 | if call(foo, [a, b
2 |     def bar():
  |     ^^^ Syntax Error: Expected ']', found 'def'
3 |         pass
  |
```

Reply:
Should we set `self.nesting = 0` in that case?

Reply:
I'm not sure how that would help. Can you expand a bit?

If we set it to 0, then the outer context (the one considering `(`) will not recover correctly as it'll not try to re-lex the token.

Reply:
I don't know if it would help with anything but I find it confusing that the lexer might return a `Newline` token or similar while `self.nesting > 0`. Or, if that's never the case, why it is necessary to run the relexing if `self.nesting > 0`, if there's never the chance that the lexed token is going to be different.

Reply:
I don't think it returns a `Newline` token if `nesting > 0`. Considering the above example, the inner list parsing (within `[`) will try to recover and re-lex it as a `NonLogicalNewline` and then the outer list parsing (within `(`) will re-lex it as a `Newline` because only now the nesting is 0.